Issue8 date format should be iso8601

Benrnz opened the pull request:
Fixes #8 
By default serialisation will format dates using ISO8601 format. For example
2015-12-30T23:50:51Z (UTC Date and time)
2015-12-30T15:50:51.011 (Local Date and time)

Using the standard .NET 4+ format "o" the returns trailing zeros padded out to 7 digits.
2015-12-30T23:50:51.0000000Z
However the System.Xaml.dll library was built prior and it does not include the trailing zeros.  So to match the behaviour of this I've added an extra function (RemoveUnnecessaryZeros) in DateTimeConverter.cs.

Let me know if there's anything else in there you'd like me to change.

Benrnz commented:
Hey @cwensley,
I just noticed a couple of issues with this.  I can do a better job of the regular expressions, and I need another test to verify dates with no time that end in zero.

Benrnz commented:
I've added a new commit 5c01b32 that has simplified the removal of trailing zeros with a regex.  Also made sure (with a permanent test) that dates without a time that end in 0 will not be subject to trailing zero removal.
Cheers

cwensley commented:
@Benrnz, this looks really great!  I was going to mention about the multiple Regex patterns.. (;

A few things I think we can still do to optimize a bit, I'll comment inline..

cwensley commented on src/Portable.Xaml/Portable.Xaml.ComponentModel/DateTimeConverter.cs:
```diff
@@ -45,8 +58,28 @@ public override object ConvertTo (ITypeDescriptorContext context, CultureInfo cu
+            if (kind == DateTimeKind.Utc)
+            {
+                if (UtcTidyUpPattern.IsMatch(dateFormatted))
+                    conciseDate = UtcTidyUpPattern.Replace(dateFormatted, "Z");
```
no need for Match + Replace, as it will then parse the regex twice.  Just simply doing the Replace will work here (and for the LocalTidyUpPatterns).

cwensley commented on src/Portable.Xaml/Portable.Xaml.ComponentModel/DateTimeConverter.cs:
```diff
@@ -45,8 +58,28 @@ public override object ConvertTo (ITypeDescriptorContext context, CultureInfo cu
+                if (UtcTidyUpPattern.IsMatch(dateFormatted))
+                    conciseDate = UtcTidyUpPattern.Replace(dateFormatted, "Z");
+                if (conciseDate.EndsWith(".Z"))
+                    conciseDate = conciseDate.Replace(".Z", "Z");
```
Maybe this should be `conciseDate = conciseDate.Remove(consiceDate.Length - 2, 1)` to be more efficient and to make sure it doesn't replace anything else in the string (though the odds for this one are very low)

cwensley commented on src/Portable.Xaml/Portable.Xaml.ComponentModel/DateTimeConverter.cs:
```diff
@@ -45,8 +58,28 @@ public override object ConvertTo (ITypeDescriptorContext context, CultureInfo cu
+                if (LocalTidyUpPatterns.IsMatch(dateFormatted))
+                    conciseDate = LocalTidyUpPatterns.Replace(dateFormatted, "");
+                if (conciseDate.EndsWith("."))
+                    conciseDate = conciseDate.Replace(".", string.Empty);
```
This should be `conciseDate = conciseDate.Substring(0, consiceDate.Length - 1)` for efficiency and to ensure it doesn't replace any other dots in the string.

Benrnz commented:
Hey @cwensley 
All changes done as requested.  
Also just wanted to check to make sure you had a look at the NullableConverter change, I've added a new override there to call down into the underlying generic parameter's type converter.
Cheers

cwensley commented:
@Benrnz, thanks for the changes, it looks perfect!  I did look at the NullableConverter and was about to comment on it asking why, but then I saw the tests and it made sense to me.  Thanks for checking with me, it is appreciated!

So, only a few final things and this will be ready for merging:
- Change to using tab indents.  I will be adding an [.editorconfig](http://editorconfig.org) so this can be configured in VS/XS automatically.
- A rebase/squash would be nice (if you are familiar with the procedure), but not mandatory.

Thanks again for the hard work!

Benrnz commented:
Hey Curtis,
Thanks for your patience. All spaces should be converted to tabs now.  I found a few more after squashing unfortunately.  Let me know if I've done the squash correctly, I haven't tried that before.

Cheers

cwensley commented:
@Benrnz no problem at all.  The diffs now look really good!  

I don't think the squash quite went right (it appears you did a merge, not a squash).  [Here](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes) is a great rundown of how to do it.

Note that you have to force push (`git push -f`) the branch to github after squashing as it will rewrite history.

If you still have problems, no worries.  It just makes for a much cleaner history to avoid confusion down the line.

Michael-DST commented:
OMGZ!!!  FINALLY a project that uses tabs!!! I LOVE YOU FOREVER AND EVER @cwensley !!! :) :) :)

Seriously seems like all the projects I follow are space wh0rze and tabs are left for the newbs.

Of course, I would like to see this whole silly argument [become a thing of the past where it belongs](http://blog.developers.win/2015/10/roslyn-powered-code-views/) (shameless plug: [please vote](http://visualstudio.uservoice.com/forums/121579-visual-studio/suggestions/10020390-enable-roslyn-powered-code-views)).

Anyways, great to see the work done on this project.  FWIW, I made mention of this project on a [recent Visual Studio User Voice vote](http://visualstudio.uservoice.com/forums/121579-visual-studio-2015/suggestions/11234259-open-source-system-xaml). :sunglasses:

Benrnz commented:
Hey @cwensley,
I've read through that guideline, but following those instructions I'm getting an error:

`error: could not apply 4a4a3ff... Removed the IsMatch for the Regex matching and just let the replace do the work.
When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 4a4a3ff40c2f56f48d3105ddcee6309507642be6... Removed the IsMatch for the Regex matching and just let th
replace do the work.`

This is a ways outside my comfort zone with git, so I'm hesitant to start "trying things" and risk making a mess of the history.  How do you want to proceed?

cwensley commented:
@Benrnz, that's great!  I'm glad you were able to figure it out.  Thanks again for the contribution! 

@Michael-DST, yes something like that would be great though not all of us use VS exclusively so it'll still be a major pain for the foreseeable future. );
